src/python/grpcio/grpc/_cython/_cygrpc/aio/completion_queue.pyx.pxi

```diff
@@ -30,41 +40,70 @@ cdef class BaseCompletionQueue:
 cdef class PollerCompletionQueue(BaseCompletionQueue):
 
     def __cinit__(self):
+        self._loop = asyncio.get_event_loop()
         self._cq = grpc_completion_queue_create_for_next(NULL)
         self._shutdown = False
         self._poller_thread = threading.Thread(target=self._poll_wrapper, daemon=True)
         self._poller_thread.start()
 
-    cdef void _poll(self) except *:
+        self._read_socket, self._write_socket = socket.socketpair()
+        self._write_fd = self._write_socket.fileno()
+        self._loop.add_reader(self._read_socket, self._handle_events)
+
+        self._queue = cpp_event_queue()
+
+    cdef void _poll(self) nogil:
         cdef grpc_event event
         cdef CallbackContext *context
 
         while not self._shutdown:
-            with nogil:
-                event = grpc_completion_queue_next(self._cq,
-                                                   _GPR_INF_FUTURE,
-                                                   NULL)
+            event = grpc_completion_queue_next(self._cq,
+                                                _GPR_INF_FUTURE,
+                                                NULL)
 
             if event.type == GRPC_QUEUE_TIMEOUT:
-                raise AssertionError("Core should not return GRPC_QUEUE_TIMEOUT!")
+                with gil:
+                    raise AssertionError("Core should not return GRPC_QUEUE_TIMEOUT!")
             elif event.type == GRPC_QUEUE_SHUTDOWN:
                 self._shutdown = True
             else:
-                context = <CallbackContext *>event.tag
-                loop = <object>context.loop
-                loop.call_soon_threadsafe(
-                    _handle_callback_wrapper,
-                    <CallbackWrapper>context.callback_wrapper,
-                    event.success)
+                self._queue.push(event)
```

Comment:
is this thread-safe? Having the feeling that no, but I could be wrong.

Reply:
Added a mutex. I don't have enough evidence to proof either way.